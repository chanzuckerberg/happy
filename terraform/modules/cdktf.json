{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "514a6ca1-8aff-418a-86f4-7b8beea67509",
  "sendCrashReports": "true",
  "terraformProviders": ["aws@~> 4.0"],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}
