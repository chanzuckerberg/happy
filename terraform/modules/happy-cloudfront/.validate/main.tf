module "test_validate" {
  source = "../../happy-cloudfront"

  frontend = {
    domain_name = "example.com"
    zone_id     = "1234567890"
  }
  origins = [
    {
      domain_name  = "example1.com"
      path_pattern = "/api/oauth/*"
    },
    {
      domain_name  = "example2.com"
      path_pattern = "/"
    }
  ]
  tags = {
    owner     = ""
    service   = ""
    project   = ""
    env       = ""
    managedBy = ""
  }
  providers = {
    aws.useast1 = aws.useast1
  }
}

provider "aws" {
  alias = "useast1"
}