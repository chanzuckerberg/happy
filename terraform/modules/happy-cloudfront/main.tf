module "cert" {
  source = "github.com/chanzuckerberg/cztack//aws-acm-certificate?ref=v0.59.0"

  cert_domain_name    = var.frontend.domain_name
  aws_route53_zone_id = var.frontend.zone_id
  tags                = var.tags

  # NOTE: certificates need to be in us-east-1 for cloudfront
  providers = {
    aws = aws.useast1
  }
}

resource "random_pet" "this" {
  keepers = {
    origin_domain_name = var.origin.domain_name
  }
}

locals {
  origin_id = "happy_cloudfront_${random_pet.this.keepers.origin_domain_name}"
}

resource "aws_cloudfront_distribution" "this" {
  enabled     = true
  comment     = "Forward requests from ${var.frontend.domain_name} to ${var.origin.domain_name}."
  price_class = var.price_class
  aliases     = [var.frontend.domain_name]

  viewer_certificate {
    acm_certificate_arn      = module.cert.arn
    ssl_support_method       = "sni-only"
    minimum_protocol_version = "TLSv1.2_2021"
  }

  origin {
    domain_name = var.origin.domain_name
    origin_id   = local.origin_id
    custom_origin_config {
      https_port             = "443"
      origin_protocol_policy = "https-only"
      origin_ssl_protocols   = ["TLSv1.2"]
    }
  }

  default_cache_behavior {
    viewer_protocol_policy   = "https"
    target_origin_id         = local.origin_id
    allowed_methods          = var.allowed_methods
    cached_methods           = var.allowed_methods
    origin_request_policy_id = var.origin_request_policy_id
    cache_policy_id          = var.cache_policy_id

    min_ttl     = var.cache.min_ttl
    default_ttl = var.cache.default_ttl
    max_ttl     = var.cache.max_ttl
    compress    = var.cache.compress
  }

  restrictions {
    geo_restriction {
      locations        = var.geo_restriction_locations
      restriction_type = "whitelist"
    }
  }

  tags = var.tags
}

resource "aws_route53_record" "alias_ipv4" {
  zone_id = var.frontend.zone_id
  name    = var.frontend.domain_name
  type    = "A"

  alias {
    name                   = aws_cloudfront_distribution.this.domain_name
    zone_id                = aws_cloudfront_distribution.this.hosted_zone_id
    evaluate_target_health = false
  }
}

resource "aws_route53_record" "alias_ipv6" {
  zone_id = var.frontend.zone_id
  name    = var.frontend.domain_name
  type    = "AAAA"

  alias {
    name                   = aws_cloudfront_distribution.this.domain_name
    zone_id                = aws_cloudfront_distribution.this.hosted_zone_id
    evaluate_target_health = false
  }
}