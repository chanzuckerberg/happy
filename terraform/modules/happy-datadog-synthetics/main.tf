data "datadog_synthetics_locations" "locations" {}

resource "datadog_synthetics_test" "test_api" {
  for_each = local.synthetics
  type     = "api"
  subtype  = "http"
  request_definition {
    method = "GET"
    url    = var.synthetic_url
  }
  assertion {
    type     = "statusCode"
    operator = "is"
    target   = "200"
  }
  locations = keys(data.datadog_synthetics_locations.locations.locations)
  options_list {
    tick_every = 900

    retry {
      count    = 2
      interval = 300
    }

    monitor_options {
      renotify_interval = 120
    }
  }
  name    = "A website synthetic for the happy stack ${var.deployment_stage} ${var.stack_name} ${var.service_name} located at ${var.synthetic_url}"
  message = "Notify @opsgenie-${var.opsgenie_owner}"
  status  = "live"
  tags    = var.tags
}
