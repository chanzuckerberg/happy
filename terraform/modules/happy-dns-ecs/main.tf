# This template creates a route53 cname for a shared alb resource.
#

locals {
  dns_prefix = var.custom_stack_name == var.app_name ? var.app_name : "${var.custom_stack_name}-${var.app_name}"
}

data aws_route53_zone dns_record {
  name = var.zone
}

resource aws_route53_record dns_record_0 {
  name    = "${local.dns_prefix}.${var.zone}"
  type    = "A"
  zone_id = data.aws_route53_zone.dns_record.zone_id

  alias {
    name                   = var.alb_dns
    zone_id                = var.canonical_hosted_zone
    evaluate_target_health = false
  }
  tags = var.tags
}
