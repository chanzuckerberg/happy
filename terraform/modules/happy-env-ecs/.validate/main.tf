module "test_validate" {
  source    = "../../happy-env-ecs"
  name      = "test"
  base_zone = "test"
  cloud-env = {
    database_subnet_group = "test"
    database_subnets      = ["test"]
    private_subnets       = ["test"]
    public_subnets        = ["test"]
    vpc_cidr_block        = "test"
    vpc_id                = "test"
  }
  tags = {
    env       = "test"
    managedBy = "teste"
    owner     = "test"
    project   = "test"
    service   = "test"
  }
  providers = {
    aws.czi-si = aws.czi-si
  }
}

provider "aws" {
  alias = "czi-si"
}