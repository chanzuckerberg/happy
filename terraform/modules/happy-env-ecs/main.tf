locals {
  env     = var.tags["env"]
  owner   = var.tags["owner"]
  project = var.tags["project"]
  service = var.tags["service"]
}
