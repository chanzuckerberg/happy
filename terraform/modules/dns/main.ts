import { Construct } from "constructs";
import { App, TerraformStack, TerraformVariable } from "cdktf";
import { DataAwsRoute53Zone } from "@cdktf/provider-aws/lib/data-aws-route53-zone";
import { Route53Record } from "@cdktf/provider-aws/lib/route53-record";
import { DataAwsAlb } from "@cdktf/provider-aws/lib/data-aws-alb"
import { AwsProvider } from "@cdktf/provider-aws/lib/provider";

export class HappyDNS extends Construct {
  record: Route53Record

  constructor(scope: Construct, id: string, albName: string, albZoneId: string, recordName: string, zoneId: string) {
    super(scope, id)
    this.record = new Route53Record(this, "happy_stack_record", {
      name: recordName,
      zoneId: zoneId,
      type: "A",
      alias: [{
        name: albName,
        zoneId: albZoneId,
        evaluateTargetHealth: false,
      }]
    })
  }
}

export class HappyDNSModule extends TerraformStack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const recordName = new TerraformVariable(this, "record_name", {
      type: "string",
      description: "The name of the record to create"
    })
    const zoneName = new TerraformVariable(this, "zone_name", {
      type: "string",
      description: "The zone ID of the record to attach to",
    });
    const albARN = new TerraformVariable(this, "alb_arn", {
      type: "string",
      description: "The ARN of the ALB to use for this happy stack",
    });

    const alb = new DataAwsAlb(this, "alb", {
      arn: albARN.stringValue,
    })

    const zone = new DataAwsRoute53Zone(this, "zone", {
      name: zoneName.stringValue
    })

    new HappyDNS(this, "happy-dns", alb.name, alb.zoneId, recordName.stringValue, zone.zoneId)
  }
}

const app = new App();
const stack = new HappyDNSModule(app, "happy-dns");
new AwsProvider(stack, "AWS", {
  region: "us-west-2",
});
app.synth();
