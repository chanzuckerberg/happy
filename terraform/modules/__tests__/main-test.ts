// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import "cdktf/lib/testing/adapters/jest"; // Load types for expect matchers
import { TerraformStack, Testing } from "cdktf";
import { HappyDNS } from "../main"
import { Route53Record } from "@cdktf/provider-aws/lib/route53-record";
describe("happy DNS", () => {
  it("create an AWS Route53 record", () => {
    const construct = Testing.synthScope((scope) => {
      new HappyDNS(scope, "my-dns", "albName", "albZoneId", "zone.czi.technology", "zoneId", "stackName", "appName");
    })
    expect(construct).toHaveResourceWithProperties(Route53Record, {
      type: "A",
      name: "stackName-appName.zone.czi.technology",
      zone_id: "zoneId", // weird, it changes from camelCase to snake_case otherwise the test will fail
      alias: [{
        name: "albName",
        zone_id: "albZoneId",
        evaluate_target_health: false,
      }]
    });
  });

  it("it looks like it did before", () => {
    const app = Testing.app();
    const stack = new TerraformStack(app, "test-happy-dns");
    new HappyDNS(stack, "my-dns", "albName", "albZoneId", "zone.czi.technology", "zoneId", "stackName", "appName");
    expect(Testing.synth(stack)).toMatchSnapshot();
  });


  //   it("Tests a combination of resources", () => {
  //     expect(
  //       Testing.synthScope((stack) => {
  //         new TestDataSource(stack, "test-data-source", {
  //           name: "foo",
  //         });

  //         new TestResource(stack, "test-resource", {
  //           name: "bar",
  //         });
  //       })
  //     ).toMatchInlineSnapshot();
  //   });
  // });

  // describe("Checking validity", () => {
  //   it("check if the produced terraform configuration is valid", () => {
  //     const app = Testing.app();
  //     const stack = new TerraformStack(app, "test");

  //     new TestDataSource(stack, "test-data-source", {
  //       name: "foo",
  //     });

  //     new TestResource(stack, "test-resource", {
  //       name: "bar",
  //     });
  //     expect(Testing.fullSynth(app)).toBeValidTerraform();
  //   });

  //   it("check if this can be planned", () => {
  //     const app = Testing.app();
  //     const stack = new TerraformStack(app, "test");

  //     new TestDataSource(stack, "test-data-source", {
  //       name: "foo",
  //     });

  //     new TestResource(stack, "test-resource", {
  //       name: "bar",
  //     });
  //     expect(Testing.fullSynth(app)).toPlanSuccessfully();
  //   });
  // });
});
