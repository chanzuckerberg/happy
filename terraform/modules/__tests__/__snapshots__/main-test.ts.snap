// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy DNS it looks like it did before 1`] = `
"{
  "provider": {
    "aws": [
      {
        "region": "us-west-2"
      }
    ]
  },
  "resource": {
    "aws_route53_record": {
      "my-dns_happy_stack_record_780D0870": {
        "alias": [
          {
            "evaluate_target_health": false,
            "name": "albName",
            "zone_id": "albZoneId"
          }
        ],
        "name": "stackName-appName.zone.czi.technology",
        "type": "A",
        "zone_id": "zoneId"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "4.36.1"
      }
    }
  }
}"
`;
